great_expectations/zep/bi_directional_dict.py

```diff
@@ -0,0 +1,62 @@
+import pprint
+from collections import UserDict
+from typing import Hashable, Iterable, Mapping, Optional, TypeVar
+
+T = TypeVar("T", bound=Hashable)
+
+
+class BiDict(
+    UserDict,
+    Mapping[T, T],
+):
+    """
+    Dict-like Mapping object that creates keys from values and values from keys.
+    Because of this, all values must be Hashable.
+
+    If a Mapping-like object is passed as the first parameter, its key/values will be
+    unpacked (and combined with kwargs) into the new `BiDict` object.
+    """
+
+    def __init__(
+        self,
+        __dict: Optional[Mapping[Hashable, Hashable]] = None,
+        **kwargs: Hashable,
+    ):
+        if isinstance(__dict, Mapping):
+            kwargs: dict = {**__dict, **kwargs}  # type: ignore[no-redef]
+            __dict = None
+        super().__init__(__dict, **kwargs)
+
+    def __delitem__(self, key: Hashable):
+        value = self.data.pop(key)
+        super().pop(value, None)
+
+    def __setitem__(self, key: Hashable, value: Hashable):
+        if key in self:
+            del self[self[key]]
+        if value in self:
+            del self[value]
+        super().__setitem__(key, value)
+        super().__setitem__(value, key)
+
+    def __repr__(self) -> str:
+        return f"{type(self).__name__}({super().__repr__()})"
+
+    def __str__(self) -> str:
+        return str(self.data)
+
+    def contains_anyof(self, collection_: Iterable[Hashable]) -> bool:
+        return bool(set(collection_).intersection(self.keys()))
+
+
+# This makes BiDict pretty printable like a regular dict
+if isinstance(getattr(pprint.PrettyPrinter, "_dispatch"), dict):
```

Comment:
Being able to pretty print this structure is definitely not necessary 😄 , especially for this particular data structure because by the nature of it only accepting hashable values it's unlikely to be deeply nested.

But it does need this in order to be pretty-printable.

```python
import string
d = {
    "alpha": {},
    "bravo": {
        "red": "RED",
        "blue": BiDict(foo="bar", letters=tuple(string.ascii_uppercase[:10])),
    },
}
```

```python
# without PP impl
{'alpha': {},
 'bravo': {'blue': BiDict({'foo': 'bar', 'bar': 'foo', 'letters': ('A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'), ('A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'): 'letters'}),
           'red': 'RED'}}

# with PP impl
{'alpha': {},
 'bravo': {'blue': {'bar': 'foo',
                    'foo': 'bar',
                    'letters': ('A',
                                'B',
                                'C',
                                'D',
                                'E',
                                'F',
                                'G',
                                'H',
                                'I',
                                'J'),
                    ('A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J'): 'letters'},
           'red': 'RED'}}
   ```

Reply:
I took this out.